### PR TITLE
Fix expandOnBoot=false and rigid path to nix-path-registration

### PR DESCRIPTION
**Pull request by rcambrj**
## Description of changes

* fix bug where `sdImage.expandOnBoot=false` would prevent nix-store from loading DB
* fix bug where nix store DB would not be loadable if root was not mounted on `/`

## Further context

I run my machines with impermanence, eg:

```
  fileSystems = mkForce {
    "/" = {
      fsType = "tmpfs";
    };
    [...]
};
```

This means that when using `sd-image.nix`, I don't use `expandOnBoot`, because this attempts to expand whatever partition is mounted at `/`. A further improvement would be to specify which partition to `expandOnBoot` - I may contribute that improvement after this.

As a result of using `expandOnBoot=false`, with `nix.gc.automatic=true`, the system would come up for a few seconds only to begin deleting all of its store, thereby destroying itself as the machine considered itself garbage. Without automatic garbage collection, the machine would come up and stay up, but was unable to successfully `nixos-rebuild switch`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/installer/sd-card/sd-image.nix [hunk 1]
```diff
@@ -255,11 +264,8 @@ in
       '';
     }) {};
 
-    boot.postBootCommands = lib.mkIf config.sdImage.expandOnBoot ''
-      # On the first boot do some maintenance tasks
-      if [ -f /nix-path-registration ]; then
-        set -euo pipefail
-        set -x
+    boot.postBootCommands = let
+      expandOnBoot = if config.sdImage.expandOnBoot then ''
```
Comment (zimbatm) [inline nixos/modules/installer/sd-card/sd-image.nix:268]:
```suggestion
      expandOnBoot = lib.optionalString config.sdImage.expandOnBoot ''
```

File: nixos/modules/installer/sd-card/sd-image.nix [hunk 2]
```diff
@@ -269,16 +275,25 @@ in
         echo ",+," | sfdisk -N$partNum --no-reread $bootDevice
         ${pkgs.parted}/bin/partprobe
         ${pkgs.e2fsprogs}/bin/resize2fs $rootPart
+      '' else "";
```
Comment (zimbatm) [inline nixos/modules/installer/sd-card/sd-image.nix:278]:
```suggestion
      '';
```

**Review by zimbatm — COMMENTED** [review]
Looks good overall

File: nixos/modules/installer/sd-card/sd-image.nix (hunk 1 above)
Comment (rcambrj) [reply nixos/modules/installer/sd-card/sd-image.nix]: thanks for the review :) fixed and fixed up

File: nixos/modules/installer/sd-card/sd-image.nix (hunk 2 above)
Comment (rcambrj) [reply nixos/modules/installer/sd-card/sd-image.nix]: fixed and fixed up

**Comment by alyssais** [comment]
Impermanence still has to persist certain paths, right?  So why not just make /nix-path-registration one of those paths?

**Comment by rcambrj** [comment]
@alyssais that could work, I haven't tested it but I don't see why it wouldn't. I think it'd be prudent to add some documentation somewhere for `sdImage` to indicate that this needs to be done. Would you agree? I can't see a particularly visible spot for this though, since there's no "main" `sdImage.enable` to put caveats.

**Comment by alyssais** [comment]
Would it make sense for it to be documented along with other impermanence things?

**Comment by rcambrj** [comment]
@alyssais are you suggesting [the wiki](https://nixos.wiki/wiki/Impermanence)? I don't see a section on using `sd-image.nix` there. Is there somewhere else I've missed?

**Comment by alyssais** [comment]
I'm not suggesting anywhere in particular, because I don't know where people go to find documentation on impermanence.  Given that you're using this and I'm not, and you presumably learned about it from somewhere, I think you'll have a better idea than me where the most helpful place to document this would be.

**Comment by rcambrj** [comment]
@alyssais I appreciate you taking the time to repeatedly help my smooth brain to figure out how to create a well-shaped change request 😆 I'm still very much in my early stages learning nix, nixos and impermanence. My infinitely inexperienced mind would first go to the module's documentation to find out any gotchas/caveats. As you can see this is my first contribution, so I'm happy to be guided in the right direction.

**Comment by alyssais** [comment]
I've asked in the [Impermanance matrix room](https://matrix.to/#/%23impermanence:nixos.org).  Maybe somebody will know better than us. :)

**Comment by zimbatm** [comment]
I did a triple-read on the code, and it looks good.

**Comment by alyssais** [comment]
Huh? We had agreed not to do it this way, and instead document it.

**Comment by zimbatm** [comment]
Sorry for the misunderstanding, @alyssais. My impression was that you were seeking external advice. I must have missed the documentation consensus. I still think the code change is acceptable and makes it a bit more modular.

**Pull request merged**